### PR TITLE
PARQUET-2371: Resolve japicmp CI failure

### PR DESCRIPTION
**Pull request by amousavigourabi**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
        No user facing changes, only resolves CI failure.

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

---

Japicmp failed on newly introduced classes, during review this did not cause a failure. This likely has something to do with the japicmp 0.18.1 update we pushed since then (for this, see #1176). Includes bumping japicmp to 0.18.2.
The false positives were all under the banner of `CLASS_GENERIC_TEMPLATE_CHANGED`. Similar problems seem to have been reported before at siom79/japicmp#345.

### CODE REVIEW

**Comment by amousavigourabi** [comment]
@ConeyLiu @wgtmac, this is a patch for the problems we discussed at #1141.

File: pom.xml [hunk 1]
```diff
@@ -68,7 +68,7 @@
     <jackson.package>com.fasterxml.jackson</jackson.package>
     <jackson.version>2.15.2</jackson.version>
     <jackson-databind.version>2.15.2</jackson-databind.version>
-    <japicmp.version>0.18.1</japicmp.version>
+    <japicmp.version>0.18.2</japicmp.version>
```
Comment (ConeyLiu) [inline pom.xml:71]: After reverting the version back to 0.16.0, the problem is gone.

**Review by ConeyLiu — APPROVED** [review]
I am fine with these changes if we want to keep the `japicmp` version. Or else we should revert the version change.

**Review by wgtmac — APPROVED** [review]
+1 on my side.

@Fokko @gszadovszky Do you have any comment?

File: pom.xml [hunk 2]
```diff
@@ -546,6 +546,13 @@
               <!-- japicmp is overly aggressive on interface types in signatures, a type was changed to a supertype but this still triggers it -->
               <exclude>org.apache.parquet.hadoop.ColumnChunkPageWriteStore</exclude>
               <exclude>org.apache.parquet.hadoop.ParquetRecordWriter</exclude>
+              <!-- likely japicmp bug, triggers on new interface methods after updating to 0.18.1 -->
+              <exclude>org.apache.parquet.conf.ParquetConfiguration#getClass(java.lang.String,java.lang.Class,java.lang.Class)</exclude>
+              <exclude>org.apache.parquet.hadoop.util.SerializationUtil#readObjectFromConfAsBase64(java.lang.String,org.apache.parquet.conf.ParquetConfiguration)</exclude>
+              <exclude>org.apache.parquet.conf.HadoopParquetConfiguration#getClass(java.lang.String,java.lang.Class,java.lang.Class)</exclude>
+              <exclude>org.apache.parquet.avro.AvroParquetReader#builder(org.apache.parquet.io.InputFile,org.apache.parquet.conf.ParquetConfiguration)</exclude>
+              <exclude>org.apache.parquet.util.Reflection#newInstance(java.lang.Class)</exclude>
```
Comment (wgtmac) [inline pom.xml:554]: This class is removed in this PR. Should this line be deleted as well?

Comment (wgtmac) [inline pom.xml:554]:
```suggestion
```

**Comment by wgtmac** [comment]
I'll merge this when the CI is green to unblock CIs of other PRs.

**Comment by gszadovszky** [comment]
Sorry @wgtmac, @ConeyLiu, for the late response.
I don't think it is a bug in japicmp to fail with a new method in an existing interface. It would fail a 3rd party if it implements the interface. To avoid keep excluding files we might add a default implementation for the new methods instead, if possible.

**Comment by amousavigourabi** [comment]
@gszadovszky, the issue at hand was not that adding methods to already existing interfaces caused failures. As you noted, this would be perfectly fine and expected behaviour. Instead, japicmp failed on wholly new interfaces (see `org.apache.parquet.conf.ParquetConfiguration`), and new method implementations (see `org.apache.parquet.hadoop.thrift.TBaseWriteSupport`), under quite a vague motivation `CLASS_GENERIC_TEMPLATE_CHANGED`.

**Comment by gszadovszky** [comment]
Thanks for the clarification, @amousavigourabi. We cannot do anything but excluding these classes, then, I'm afraid. But the comment is a bit misleading.
Have you tried creating a bug for japicmp about it?

**Pull request merged**